ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -1550,6 +1588,62 @@ static void LogError(const char* formatString, ...)
     vasprintf(&logString, formatString, dataArgs);
     va_end(dataArgs);
     
-    s_callbacks.LogError(logString);
+    if (s_callbacks.LogError != nullptr)
+    {
+       s_callbacks.LogError(logString);
+    }
+    else
+    {
+        #ifdef DEBUG
+        cout << "LogError: s_callbacks.Info is [NULL], Unable to log: %s " << logString << endl;
+        #endif
+    }
+    
+    free(logString);
+}
+
+static void LogWarning(const char* formatString, ...)
+{
+    va_list dataArgs;
+    char* logString = nullptr;
+
+    va_start(dataArgs, formatString);
+    vasprintf(&logString, formatString, dataArgs);
+    va_end(dataArgs);
+ 
+    if (s_callbacks.LogWarning != nullptr)
+    {
+       s_callbacks.LogWarning(logString);
+    }
+    else
+    {
+        #ifdef DEBUG
+        cout << "LogWarning: s_callbacks.LogWarning is [NULL], Unable to log: %s " << logString << endl;
+        #endif
+    }
+    
+    free(logString);
+}
+
+static void LogInfo(const char* formatString, ...)
```

Comment:
These 3 functions, `LogInfo`, `LogWarning`, and `LogError` are extremely similar. Surely we can put just about everything (apart from the `va_start` and `va_end` calls) into a helper function?

Reply:
On the GVFS side we did separate these out (RelatedWarning/RelatedInfo/RelatedError), so I kept that consistent.  

We could switch a general function Log(ErrorLevel level, const char* formatString, …)

I err'd on the side of consistency here.  I think the boilerplate is straightforward, and I also totally see why you called this out.

@wilbaker

Reply:
I think it makes sense to have a common helper function (naming suggestion: `LogEvent`) that can be called by all three `Log` functions and takes an `EventLevel` parameter.  

We could also have the provider implement a single `LogEvent` callback rather than implementing 3.

Reply:
> I err'd on the side of consistency here. I think the boilerplate is straightforward, and I also totally see why you called this out.

I don't know, 3 instances seems a good threshold to start applying DRY. And even if the backend of where this logging all ends up does repeat itself, we can plug into that with a simple `switch` block or whatever.

And in terms of consistency, the Kext's `KextLog_Error`, `KextLog_Info`, and `KextLog` are implemented as macros which expand to calls to `KextLog_Printf()` with `KEXTLOG_ERROR`, `KEXTLOG_INFO`, and `KEXTLOG_DEFAULT` as the first argument. So the user space native code would be consistent with that at least.

Reply:
I discussed this offline with @jeschu1 and agreed that we can leave this as-is for now.  There's not currently a lot of code duplication that can be removed with a common helper, and we can always revisit this later if it grows any larger.